rakelib/upload.rake

```diff
@@ -0,0 +1,84 @@
+require 'aws-sdk'
+require 'json'
+require 'parallel'
+require 'mime/types'
+
+desc "upload files to s3 differentially"
+task :upload_to_s3 do
+
+  S3_BUCKET = ENV['S3_BUCKET']
+  REMOTE_NAME = ENV['REMOTE_NAME'] || "origin"
+
+  def compute_md5_of_local_object(local_file)
+    return %Q{"#{Digest::MD5.file(local_file).to_s}"}
+  end
+
+  def has_md5_changed?(etag_from_s3, computed_md5)
+    return !etag_from_s3.eql?(computed_md5)
+  end
+
+  s3_client = Aws::S3::Client.new(region: 'us-east-1')
+  objects = s3_client.list_objects(bucket: S3_BUCKET)
+  objects_from_s3 = {}
+  objects.contents.each do |object|
+    objects_from_s3[object.key] = object.etag
+  end
+
+  repo_url = `git config --get remote.#{REMOTE_NAME}.url`.strip
+  # Whatever's there on gh-pages is what we need to upload to s3
+  rm_rf "build"
+  sh("git clone #{repo_url} build --branch gh-pages --depth 1")
+
+  cd "build" do
+    rm_rf ".git"
+    # each key represents a path on the local file relative to build directory if it exists, else it needs to be deleted from the bucket
+    need_to_be_deleted = []
+    need_to_be_changed = []
+    objects_from_s3.keys.each do |key|
+      if File.exist?(key)
+        if has_md5_changed?(objects_from_s3[key], compute_md5_of_local_object(key))
+          need_to_be_changed << key
+        end
+      else
+        need_to_be_deleted << {key: key}
+      end
+    end
+
+    local_files =  Dir.glob('**/*', File::FNM_DOTMATCH).reject {|fn| File.directory?(fn) }
+    need_to_be_created = local_files.select {|object_to_be_created| !objects_from_s3.include?(object_to_be_created);}
+
+    puts "Creating..."
+    Parallel.map(need_to_be_created, in_threads: 5) do |file|
+      puts "Uploading new file #{file} to #{S3_BUCKET}/#{file}"
+      s3_client.put_object({acl: "public-read",
+                           body: File.read(file),
+                           bucket: S3_BUCKET,
+                           cache_control: "max-age=600",
+                           content_type: MIME::Types.type_for(file).first.content_type,
+                           content_md5: Digest::MD5.file(file).base64digest,
+                           key: file
+                          })
+    end
+
+    puts "Syncing changed files..."
+    Parallel.map(need_to_be_changed, in_threads: 5) do |file|
+      puts "Uploading changed file #{file} to #{S3_BUCKET}/#{file}"
+      s3_client.put_object({acl: "public-read",
+                           body: File.read(file),
+                           bucket: S3_BUCKET,
+                           cache_control: "max-age=600",
+                           content_type: MIME::Types.type_for(file).first.content_type,
+                           content_md5: Digest::MD5.file(file).base64digest,
+                           key: file
+                          })
+
+    end
+
+    puts "Deleting files that don't exist locally..."
```

Comment:
Its one request to delete all files. Does it make sense to have a loop just to print the file names?